### DISCUSSION TITLE
Kvantum as first-class citizen in LXQt

### DISCUSSION THREAD

**Discussion by AzumaHazuki** [lxqt/lxqt#2570] [Ideas]
I'd been thinking about this one for several months now: given that @tsujan created Kvantum in the first place and that he's one of the active developers here, maybe we should have it as part of the base LXQt set? Qt6 to my knowledge doesn't have an equivalent to Qt5's styleplugins package, so it only comes with Fusion, which acts a little weird with some themes and color schemes. Kvantum also seems to handle transparency properly everywhere, even on plain QWidgets, where other Qt themes don't.

The problem I can see with this isn't bloating LXQt (Kvantum is surprisingly light and quick!) but _overcomplicating_ it. The simplest way to implement this would be probably be to tweak the Widget Style section; I propose renaming it to something like "App Styles" and consolidating the GTK Style page into it. Envision three dropdowns, Qt Theme, GTK2 Theme, and GTK3 (and 4?) Theme rather than just the current Qt Theme one.

The key here is that the "Qt Style" dropdown would only show Kvantum themes by default. To give users options but keep sane defaults, I'd also add a checkbox or button that would make the dropdown also show non-Kvantum Qt styles. 

The _other_ problem I can see is that, as far as I know, Kvantum doesn't have the ability to pick up the user's color palette like Fusion does (someone correct me if I'm wrong here!). If this is the case, we run the risk of "KDE-ifying" LXQt in the sense of adding more fragmentary pieces to a uniform look or becoming reliant on "Global Look and Feel" packages.

I think this idea has serious merit but would like the community's -- especially the devs'! -- feedback on it.

**Comment by tsujan** [comment]
Kvantum is made to be DE-independent and will remain so. LXQt is agnostic toward Qt widget styles and will remain so. Kvantum can be used with any DE, as LXQt can use any Qt style.

> The other problem I can see is that, as far as I know, Kvantum doesn't have the ability to pick up the user's color palette

That's not a problem in Kvantum but a part of its design: Kvantum is an **SVG-based** theme engine, and its themes should have their own colors compatible with their SVG files.

**Comment by AzumaHazuki** [comment]
You're missing the point, and I'm beginning to suspect you're missing it deliberately every time I bring up the fact that LXQt has serious presentation and planning issues.

Yes, LXQt is WM/Compositor-agnostic...but virtually every user of it uses it with Openbox by default. Yes, LXQt is Qt-theme agnostic...but virtually every user of it uses it with the Fusion style by default. Yes, Kvantum is not exclusive to any one DE, **but I wasn't suggesting we make it so in the first place.**

The idea of sane defaults is not a GNOME-ism (no, their problem is _in_sane defaults...). There is a huge difference between consistency and lack of function. I was very surprised and very happy when the devs defaulted to Clearlooks for fresh vanilla installs, given how frankly ugly and unfinished-looking most of the included themes are.

So what would be another proposal then, QtCurve? An entirely new Qt style or two for LXQt itself that could read the active color palette?

**Reply by stefonarch** [reply]
Theming and choosing the Wm is basically a distro's job (debian ships xfwm4) - afaik only arch ships our defaults and maybe you never saw a vanilla installation of Xfce ;)

> ..but virtually every user of it uses it with Openbox

![immagine](https://github.com/lxqt/lxqt/assets/10681413/02cc877d-a2dc-4254-b141-fc029641818e)

**Reply by stefonarch** [reply]
btw thanks  for the fish...

>given how frankly ugly and unfinished-looking most of the included themes are.

**Reply by Chiitoo** [reply]
We tend to follow upstreams on Gentoo Linux in general as closely as we can.

Window Manager is a bit tricky, since it may mean building a lot of additional packages.

For ages past, we have had a USE-flag option for pulling in Openbox via the "meta" package that pulls in all of LXQt, according to other user choices, but I'm removing that since it has the effect of making it seem like Openbox is a part of LXQt, and, well, it has been quite dormant for a decade or so...

So, going forward, instead of suggesting anything like that at least as long as there is no upstream default, users will see a message about installing whichever Window Manager they like.

**Reply by tsujan** [reply]
> users will see a message about installing whichever Window Manager they like.

I don't know about Gentoo users, but I think most users wouldn't like that — many users may even not know what a "Window Manager" is ;)

So, I think it's good practice that distros include a WM of their choice in their LXQt meta-package.

**Reply by Chiitoo** [reply]
Yeah, "sane defaults" are good to have of course, but we've already had this one behind a USE-flag so it shouldn't be a huge change.

That said, I am still considering changing Openbox to something that is more active-like instead, and doesn't require GTK or a bunch of KDE things.

**Reply by tsujan** [reply]
> I am still considering changing Openbox to something that is more active-like instead, and doesn't require GTK or a bunch of KDE things.

There should be some. Please tell us which one you choose.

Among full-fledged X11 WMs that come with their compositors, I prefer KWin, but as you said, it has several KDE dependencies. Xfwm4 may have fewer dependencies. Although it needs GTK3, GTK has never been a taboo to LXQt ;)

With Wayland, display server = WM = compositor. However, it's too soon to introduce Wayland to average users — advanced users will find their way, e.g., with the help of @stefonarch's scripts.

**Reply by sfalken** [reply]
For openSUSE, as long as Wayland isn't a viable option, we ship openbox as the WM, when installing the LXQt pattern, and that's not likely to change, our userbase has been pretty vocal that kwin isn't acceptable as a default, because it's "bloated" (their words, not mine, as I don't agree) due to dependencies in the way it's packaged.

Obviously that isn't LXQt upstreams problem, or concern, but frankly it would be nice to see some sort of reference implementation from upstream, even if upstream isn't interested in declaring anything a "default"

**Reply by stefonarch** [reply]
Every choice of the 3 main WM in X11 has its downsides, but I think we can soon propose a wayland session where there will be more choice (at least 5, with only kwin with the same issues for it's settings). 
If we really should declare a default WM in future it would be labwc probably.

**Reply by tsujan** [reply]
Alf's choices in Debian were really good. I remember that I installed Debian's LXQt on VirtualBox when Alf was here, and it was very neat. He'd chosen Xfwm4 as the WM.

If there isn't an X11 WM as good as what Openbox was when it was alive, Xfwm4 may be the best choice.

> ... as Wayland isn't a viable option...

Unlike X11, with Wayland, we might recommend LabWC because of its stability, active and careful development, high performance, and the responsiveness of its developers. Strangely enough, it seems to me that Wayland recently provides more good options than X11.

**Reply by Chiitoo** [reply]
I'm thinking of going with KWin for Gentoo.

It is what I use myself, and it's just one USE-flag away for users to not install, and install something they prefer.

I've played around with Xfwm, Fluxbox, Awesome, and IceWM at least, which all seem like good options, but they all also come with their own panel (and maybe more), which makes things awkward at least for a souirce-based distro at least.

Some kind of a reference implementation indeed would be nice.  Could be useful for bug triaging too, like for instance if someone is having problems with KWin, but does it happen with lxqt-win?  Not sure how useful in that sense it would be though.

**Reply by tsujan** [reply]
> like for instance if someone is having problems with KWin

I've always used KWin with LXQt+X11 and have been very happy with it. Of course, it has its problems for some users, but which program doesn't?

The main problem with any WM may be that their shortcuts can interfere with lxqt-globalkeys. That's true for KWin too. Sadly, `kglobalacceld` is needed for some shortcuts related to changing desktop. Personally, I disable all of its other shortcuts to prevent interference with lxqt-globalkeys.

**Reply by Chiitoo** [reply]
Yeah, from Gentoo perspective, it means building a lot of dependencies when compared to Openbox for example, though it can be trimmed down a lot via certain choices the users using Gentoo have.

It's still not that much, so I think it might even be a positive surprise for some who never tried it.

**Reply by stefonarch** [reply]
I also prefer it on X11. I think you've seen https://github.com/lxqt/lxqt/wiki/ConfigWindowManagers#kwin

I didn't test the "showdesktop" issue though on x11. Also using only lxqt-globalkeys  no idea if Alt-tab is working but @tsujan  can tell. Also for grid and other exposures kwin's shortcuts are needed. The issue is or manually editing (which I did) or installing all of the plasma stuff.

**Reply by tsujan** [reply]
> Also using only lxqt-globalkeys no idea if Alt-tab is working but @tsujan can tell. 

Working under LXQt+LabWC, I can only tell from memory for now. As far as I remember, switching to next/previous workspace didn't need kglobalaccel and could be done by

```
qdbus org.kde.KWin /KWin nextDesktop
qdbus org.kde.KWin /KWin previousDesktop
```

Other actions needed kglobalaccel, although their shortcuts could be removed from it and included in lxqt-globalkey by using commands like

```
qdbus org.kde.kglobalaccel /component/kwin invokeShortcut "Switch to Next Desktop"
qdbus org.kde.kglobalaccel /component/kwin invokeShortcut "Switch to Previous Desktop"
qdbus org.kde.kglobalaccel /component/kwin invokeShortcut "Window to Next Desktop"
qdbus org.kde.kglobalaccel /component/kwin invokeShortcut "Window to Previous Desktop"
qdbus org.kde.kglobalaccel /component/kwin invokeShortcut "view_zoom_in"
qdbus org.kde.kglobalaccel /component/kwin invokeShortcut "view_actual_size"
qdbus org.kde.kglobalaccel /component/kwin invokeShortcut "MinimizeAll"
qdbus org.kde.kglobalaccel /component/kwin invokeShortcut "Window Quick Tile Left"
qdbus org.kde.kglobalaccel /component/kwin invokeShortcut "Window Quick Tile Right"
```

Please note that this list worked with KWin5, although I think it mostly works with KWin6 too.

**Reply by Chiitoo** [reply]
When switching to qt6, I did notice alt-tab not working, and after a while I noticed `kglobalacceld` wasn't running.

Never looked into what caused that change, but probably should now... I kind of thought it would have been something related to my personal set-up, which is a bit more light version of the dependencies required for KWin.

**Reply by tsujan** [reply]
> and after a while I noticed kglobalacceld wasn't running.

I saw that too. Then I realized it wasn't started by KWin anymore. So, I added it to my startup script.

**Reply by stefonarch** [reply]
I guess it's because of kwin_wayland as that doesn't use it anymore but has shortcuts built in as other compositors too. So plasma X11 will start it but LXQt needs to do it manually.

**Reply by Chiitoo** [reply]
Ah, good to know it's not just me.

I'll dig into it a bit on the KDE side too!

**Reply by AzumaHazuki** [reply]
@Chiitoo, as the (only other?) resident Gentoo user: it might make sense to have a "window-manager" USE flag that's on by default, which forces the "exactly one of/at least one of" condition for members of a group of others, with "openbox" being + in IUSE and others available but off by default? Xfwm4, Fluxbox, IceWM, KWin, etc. This would make it more obvious LXQt isn't tied to Openbox while providing a sane, familiar default.

I'm also working on distro-specific themes -- currently done with Artix and finishing up Pop and Pop Dark themes -- so am wondering if you'd like a Gentoo theme that could maybe be gated behind the "branding" use flag?

**Reply by Chiitoo** [reply]
I did consider changing 'minimal' to 'window-manager' about four years ago, but ended up leaving things as they were.

I'm not sure I'm too keen on adding even more flags, though it's certainly a good way to notify users of some of the choices out there (postinst messages for example tend to go unnoticed a lot of time I think).

Regarding Openbox specifically, I will probably be dropping it to "maintainer-needed" and if no one will pick it up, it will eventually be removed from the official repository.  There are some patches for some issues with it available, but I don't have the energy to really review them, and since they're unlikely to ever get upstream either... I'm not too keen on looking after it either.

A Gentoo theme would certainly be nice, but also not sure if I'd add another USE-flag for it.  I'll think about it in any case!

Thanks for the thoughts!

**Reply by Chiitoo** [reply]
Also regarding Openbox, since 'obconf-qt' didn't get ported to qt6, and seems like something that may not be maintained upstream any longer, it felt appropriate to remove Openbox from the ebuild as well.

That said, I did notice a PR for said port, but its future seems quite uncertain in any case.

**Reply by tsujan** [reply]
Porting/releasing of obconf-qt would give the impression that LXQt recommends Openbox. Consider it an old tool for configuring a dead app.

**Reply by AzumaHazuki** [reply]
> Also regarding Openbox, since 'obconf-qt' didn't get ported to qt6, and seems like something that may not be maintained upstream any longer, it felt appropriate to remove Openbox from the ebuild as well.
> 
> That said, I did notice a PR for said port, but its future seems quite uncertain in any case.

Regarding a potential ebuild that uses my flag ideas, here's what I've come up with (also my first time writing an ebuild with the required-use operators!)

```
# Copyright 1999-2024 Gentoo Authors
# Distributed under the terms of the GNU General Public License v2

EAPI=8

MY_PV="$(ver_cut 1-2)"

DESCRIPTION="Meta ebuild for LXQt, the Lightweight Desktop Environment"
HOMEPAGE="https://lxqt-project.org/"

if [[ ${PV} != *9999* ]]; then
	KEYWORDS="amd64 ~arm arm64 ~loong ~ppc64 ~riscv x86"
fi

LICENSE="metapackage"
SLOT="0"

IUSE="+about +admin +archiver +desktop-portal +display-manager +filemanager fluxbox
lximage +nls openbox pekwm +policykit +powermanagement processviewer screenshot 
+sddm ssh-askpass +sudo terminal +trash +windowmanager xfwm4"

REQUIRED_USE="trash? ( filemanager ) windowmanager || ( fluxbox openbox pekwm xfwm4 )"

# Note: we prefer kde-frameworks/oxygen-icons over other icon sets, as the initial
# install expects oxygen icons, until the user specifies otherwise (bug 543380)

# For lxqt-themes there is no release 1.4.0 at this time.

RDEPEND="
	=lxqt-base/lxqt-config-${MY_PV}*
	=lxqt-base/lxqt-globalkeys-${MY_PV}*
	=lxqt-base/lxqt-menu-data-${MY_PV}*
	=lxqt-base/lxqt-notificationd-${MY_PV}*
	>=lxqt-base/lxqt-panel-2.0.0
	=lxqt-base/lxqt-qtplugin-${MY_PV}*
	=lxqt-base/lxqt-runner-${MY_PV}*
	=lxqt-base/lxqt-session-${MY_PV}*
	virtual/ttf-fonts
	>=x11-themes/lxqt-themes-1.3.0
	about? ( =lxqt-base/lxqt-about-${MY_PV}* )
	admin? ( =lxqt-base/lxqt-admin-${MY_PV}* )
	archiver? ( app-arch/lxqt-archiver )
	desktop-portal? ( gui-libs/xdg-desktop-portal-lxqt )
        display-manager? (
            sddm? ( x11-misc/sddm )
            !sddm? ( x11-misc/lightdm )
        )
	filemanager? ( =x11-misc/pcmanfm-qt-${MY_PV}* )
        fluxbox? ( x11-wm/fluxbox )
        lximage? ( medix-gfx/lximage-qt )
	nls? ( dev-qt/qttranslations:6 )
        openbox? ( x11-wm/openbox )
        pekwm? ( x11-wm/pekwm )
	policykit? ( =lxqt-base/lxqt-policykit-${MY_PV}* )
	powermanagement? ( =lxqt-base/lxqt-powermanagement-${MY_PV}* )
        processviewer? ( >=x11-misc/qps-2.9.0 )
        screenshot? ( >=x11-misc/screengrab-2.8.0 )
        sddm? ( >=x11-misc/sddm-0.21.0 )
	ssh-askpass? ( =lxqt-base/lxqt-openssh-askpass-${MY_PV}* )
	sudo? ( =lxqt-base/lxqt-sudo-${MY_PV}* )
        terminal? ( >=x11-terms/qterminal-1.4.0 )
	trash? ( gnome-base/gvfs )
        xfwm4? ( xfce-base/xfwm4 )
"
```

This appears to work properly on my system. Setting the + in IUSE on the windowmanager flag but none of the actual wms themselves will force a user to pick one if they don't have one of them installed already; this will prevent that "LXQt tells me to pick a WM but it doesn't offer me anything!" problem I've seen discussed before.

**Reply by Chiitoo** [reply]
> this will prevent that "LXQt tells me to pick a WM but it doesn't offer me anything!" problem I've seen discussed before.

Yeah, that's one of the reasons I didn't want to not have /any/ WM in the "meta" package, although that kind of things kind of come with Gentoo: if you didn't install something, you won't have it, and it's working as expected.

I haven't been able to put time into pondering about what to go with in the end, but yeah, your idea may be doable.

In any case, I think this is a bit off-topic with regards to Kvantum and being a first-class citizen in LXQt, so if you would like to discuss this more, it might be best if you visit `libera/#gentoo-qt` and chat there, or send a private query to me, or e-mail me directly.  ^^

**Comment by AzumaHazuki** [comment]
The issue with LabWC is that it doesn't implement Openbox themes. I seem to recall the man dev saying he'd like to get it to 1:1 parity with Openbox on that score, but also that the way it draws decorations and such is very different from Openbox, not just "the way OB does it but in Wayland." I also seem to remember him stating it's not super-high priority at this point...so even if it gets there eventually, it could be a really long time.

Xfwm4 would be perfect for this if only it didn't have that blasted GTK dependency. Although at this point, for X11 if we're going to switch to anything it probably should be this, if for no other reason than that most people use some GTK3/4 apps already and it'll be on their system anyway...and looking at their git repo, it seems like Wayland support is already most of the way done. 

It may be best if we throw together a quick little frontend for selecting Xfwm4 themes rather than relying on Xfce's window manager settings module though, maybe integrating this into the Appearance settings submodule...

**Reply by stefonarch** [reply]
xfwm4 has an issue with desktop names, no way to use them in the pager. 
>It may be best if we throw together a quick little frontend for selecting Xfwm4 themes rather than relying on Xfce's window manager settings module though, maybe integrating this into the Appearance settings submodule...

We   stopped developing `obconf-qt` and I don't think we start duplicate existing `xfwm4-settings`.

>The issue with LabWC is that it doesn't implement Openbox themes

That's not a big issue, it implements most of them. Appearance is nice, but stability and functionality are more important.

![immagine](https://github.com/lxqt/lxqt/assets/10681413/ec55c222-1ce6-4758-b12c-c09c50ac358d)

**Reply by AzumaHazuki** [reply]
Uh, no, LabWC bloody well does _not_ render most OB themes. Try <a href=https://github.com/AzumaHazuki/lxqt-themes-nimbus/blob/main/Nimbus.tar.xz>the Nimbus theme I made</a> on LabWC and tell me if it looks anything like the screenshot. Last I checked it can't even do the SplitGradient primitive, meaning even base Openbox themes like, oh, **Clearlooks** for example? Yeah, they don't work right.

Most people are not on your level of technical competence. LXQt is also a relative newcomer and, Lubuntu aside, most distros aren't doing a heck of a lot to make it accessible or usable to a first timer. This is something that drives me up a wall with Xfce's vanilla defaults too: their initial install _looks_ like OS X but has no global menu or actual dock as the appearance would suggest to basically everyone.

You know, I don't know if it's a language barrier or what, but you and Tsujan are exceedingly difficult to work with sometimes. You come across as the human equivalents of "WONTFIX - works for my use case, no one else's matters" every time I try and suggest something to make the experience saner and less confusing.

**Reply by tsujan** [reply]
> xfwm4 has an issue with desktop names

That's just a minor issue. I installed it minutes ago (only 9.52 MiB in Arch: libwnck3, libxfce4ui, libxfce4util, xfconf and xfwm4) to test a panel PR under X11. It was light, neat and user-friendly. I'd surely use it with my X11 LXQt session if I didn't have to use KWin for reasons beyond this comment.

**Reply by Bluey26** [reply]
I could add that i have been using xfmw4 with lxqt for a period of a month (before switching to lxqt+labwc) and the only thing i did not worked for me was the minimize/maximize thing: the window was kept almost as maximized for some reason. Besides that everything seemed to work in my experience (just 1 workspace), even showdesktop and conky were working out of the box without prior modifications. A minor detail would be that i had to use the command-line commands to open `xfwm4-settings` and `xfwm4-tweaks-settings`; ~~and that the minimized windows were not counted in the appswitcher (i recall this is some kind of reported bug on their side).~~ --> Apps are present in the appswitcher but not 'previewed' unless you change to the app (there is an option which makes this kind of preview option available as you pass between the app list in the appswitcher).

**Reply by tsujan** [reply]
An important thing is that bugs can be reported and xfwm4's devs will respond. KWin also has bugs, and its devs are responsive too. But Openbox's bugs will never be fixed.

**Reply by stefonarch** [reply]
>. A minor detail would be that i had to use the command-line commands to open xfwm4-settings and xfwm4-tweaks-settings

https://github.com/lxqt/lxqt/wiki/ConfigWindowManagers#xfwm4

>You know, I don't know if it's a language barrier or what, but you and Tsujan are exceedingly difficult to work with sometimes. 

Maybe.  On the other side you really overestimate importance of theming IMO. We have heard you, and yes, labwc has no gradients (yet), take it or leave it. Some of my friends have never changed a theme, just a wallpaper is enough personalization.

**Reply by AzumaHazuki** [reply]
Then maybe it's up to me to _create_ themes for the major distros and hope their upstream is willing to use them. I already made some for Devuan ages ago, and have been working on a Mint-Y variant last night and tonight; once that's done it should be fairly easy to make a "Mint-L" version -- yes, these are going to be able to pick up on users' palette(highlight) colors.

It sounds like at a minimum I need to make them for Debian, Fedora, Gentoo, Slackware, Arch/Artix, Mageia, Void, openSuSE, Pop!_OS, Zorin OS, Manjaro, EndeavourOS, and maybe Garuda, plus simple Openbox themes that work in LabWC. Maybe this is the best way forward...

**Reply by Gnomasz** [reply]
> Then maybe it's up to me to create themes for the major distros and hope their upstream is willing to use them.

Sounds neat! Though I'd suggest to first ask the distro maintainers if they're interested so you don't lose steam on something unfulfilling. The last time I checked, the Fedora and Manjaro spins seemed to be in a bit of a disarray. Also, Manjaro already does have [it's LXQt theme](https://codeberg.org/wonky/lxqt-themes)), and so does [Debian](https://tracker.debian.org/pkg/lxqt-branding-debian) – these are the distros that I've tried.

**Reply by isf63** [reply]
I've been using Xfwm4 on X11 for about a year now and I appreciate it. The only annoyances I've had are copying the configuration desktop files with `OnlyShowIn=XFCE;` removed into `~/.local/share/applications`, and one or two occasional bugs. Actually I've encountered more bugs with Kwin, not to mention it's its dependencies.
